pkg/cluster/audiences.go

```diff
@@ -0,0 +1,57 @@
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	authentication "k8s.io/api/authentication/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	// Default value of audiences for DSCI.SM.auth.
+	defaultAudiences = []string{"https://kubernetes.default.svc"}
+)
+
+func isDefaultAudiences(specAudiences *[]string) bool {
+	return specAudiences == nil || reflect.DeepEqual(*specAudiences, defaultAudiences)
+}
+
+type TokenSupplier func() (string, error)
+
+// GetEffectiveClusterAudiences returns the audiences defined for the cluster
+// falling back to the default audiences in case of errors.
+func GetEffectiveClusterAudiences(cli client.Client, log logr.Logger, specAudiences *[]string, getTokenFunc TokenSupplier) []string {
+	if isDefaultAudiences(specAudiences) {
+		return fetchClusterAudiences(cli, log, getTokenFunc)
+	}
+	return *specAudiences
+}
+
+func fetchClusterAudiences(cli client.Client, log logr.Logger, getTokenFunc TokenSupplier) []string {
+	token, err := getTokenFunc()
+	if err != nil {
+		log.Error(err, "Error getting token, using default audiences")
+		return defaultAudiences
+	}
+
+	tokenReview := &authentication.TokenReview{
+		Spec: authentication.TokenReviewSpec{
+			Token: token,
+		},
+	}
+
+	if err = cli.Create(context.Background(), tokenReview, &client.CreateOptions{}); err != nil {
+		log.Error(err, "Error creating TokenReview, using default audiences")
+		return defaultAudiences
+	}
+
+	if tokenReview.Status.Error != "" || !tokenReview.Status.Authenticated {
+		log.Error(fmt.Errorf(tokenReview.Status.Error), "Error with token review authentication status, using default audiences")
+		return defaultAudiences
+	}
+
+	return tokenReview.Status.Audiences
```

Comment:
In the docs that I linked, there is this text over `status.audiences` of the TokenReview:

```
# If this is omitted, the token is considered to be valid to authenticate to the Kubernetes API server.
```

> Underlying AuthConfig will be defaulted if not configured specifically.
> Current implementation defaults it to "https://kubernetes.default.svc" if not specified.

IIRC, this default is on odh-model-controller side. This is OK.

My concern is that my understanding of the kubernetes docs is that the cluster may choose to omit the `status.audiences` field as an indication that the token is valid for the API Server. Yet, in that case, I see no statement that the audience would be `https://kubernetes.default.svc` in case it is omitted. Thus, we may wrongly assume it is `https://kubernetes.default.svc`.

The writing is somewhat confusing. So, I may be understanding things badly.

Reply:
https://github.com/kubernetes/kubernetes/blob/v1.31.2/pkg/apis/authentication/types.go#L83C31-L86C20
looks like we might have .status.Authenticated==true && .status.Error == "" && .status.Audiences empty case. so far i am not able to reproduce this from my rosa hcp